### ISSUE TITLE
build_mac & build_windows

### ISSUE THREAD

**Issue by nomed** [canonical/dqlite#705]
Hi dqlite team,

I am interested on building dqlite on both Windows and macOS platforms. I noticed that the canonical/raft project had some pull requests that were accepted, which included patches to support these platforms. However, it seems that after the inclusion of raft into dqlite, the necessary configurations for these platforms might be missing in the configure.ac file.

Could you please provide some information on the current status of Windows and macOS build support for dqlite? 
Are there any plans to reintroduce the patches or configurations that were previously available in canonical/raft?

Thank you for your time and assistance.

**Comment by cole-miller** [comment]
Hi @nomed, we're not currently prioritizing support for macOS and Windows. However, I can look into restoring the previous state of compatibility if that regressed during the raft merge.

**Comment by nomed** [comment]
Hi,

I appreciate your willingness to look into this. I would be happy to try to contribute and help make it work on those platforms if possible.

Thank you!

**Comment by abelcha** [comment]
I was looking to build on MacOS @nomed if you want to split the tasks

**Comment by cole-miller** [comment]
I refreshed my memory on this and I am pretty sure our build has never worked on macOS or Windows. The PRs I see to canonical/raft to support these platforms (https://github.com/canonical/raft/pull/173 and https://github.com/canonical/raft/pull/119) were never merged. I would be happy to make the build work on other platforms if it were a relatively small matter, but it is blocked by doing a big refactor of uv_os.c and uv_writer.c to isolate the code that uses KAIO, and that's a bigger task.

**Comment by petar-dambovaliev** [comment]
> Hi dqlite team,
> 
> I am interested on building dqlite on both Windows and macOS platforms. I noticed that the canonical/raft project had some pull requests that were accepted, which included patches to support these platforms. However, it seems that after the inclusion of raft into dqlite, the necessary configurations for these platforms might be missing in the configure.ac file.
> 
> Could you please provide some information on the current status of Windows and macOS build support for dqlite? Are there any plans to reintroduce the patches or configurations that were previously available in canonical/raft?
> 
> Thank you for your time and assistance.

Did you manage to find an alternative? I am in need of one now.

**Comment by marco6** [comment]
The upcoming multithreading implementation (#747) requires Linux 2.3+.

I don't think we can support Mac or Windows as well. If someone knows how to remap a pointer atomically on Windows or Mac, we can think about how to do that. For now, I will mark this as "won't fix".

**Comment by MrSparc** [comment]
I am in the same boat, I would like to integrate DQLite embedded in our projects. One of the main advantages of SQLite that made so successful was its ability to be embedded using amalgamation of its source code in C/C++ software. I was looking do the same with  DQLite. Our applications are C/C++ but cross-compiled to run on Linux (GCC), Windows (MingGW), Android (Clang), and attempt to MacOS.
I started on Mac trying to see what would be necessary to port: I had to create a wrapper over `qsort_r` (GlibC vs BSD), mutexes are using C11 `mtx` and would have to use `pthread_mutex`, semaphores `sem_t`, commenting the check for Linux Kernel  4.14, etc.
Regarding the recent development of supporting multithreading and using the `mremap` function, I don't need multithreading; I can live with the limitation of being single-threaded.

The issue is that the project uses Linux AIO, which makes porting it more complicated. I don't have much know-how but maybe they could use Libuv for this as well.

<details>

<summary>As an example, I'll leave a path of what I've found so far, in case it might be useful: (expand it)</summary>

```diff
diff --git a/configure.ac b/configure.ac
index 46d2799..ce0ddbc 100644
--- a/configure.ac
+++ b/configure.ac
@@ -56,7 +56,7 @@ AX_CODE_COVERAGE
 AC_CHECK_HEADERS([linux/io_uring.h linux/aio_abi.h])
 
 # Checks for library functions and definitions.
-AC_CHECK_DECLS(RWF_NOWAIT, [], [AC_MSG_ERROR(Linux kernel >= 4.14 required.)], [#include <linux/aio_abi.h>])
+#AC_CHECK_DECLS(RWF_NOWAIT, [], [AC_MSG_ERROR(Linux kernel >= 4.14 required.)], [#include <linux/aio_abi.h>])
 
 # Enable large file support. This is mandatory in order to interoperate with
 # libuv, which enables large file support by default, making the size of 'off_t'
diff --git a/src/roles.c b/src/roles.c
index e01f4ef..2f0f2b8 100644
--- a/src/roles.c
+++ b/src/roles.c
@@ -2,6 +2,7 @@
 
 #include "client/protocol.h"
 #include "lib/queue.h"
+#include "qsort_r_wrapper.h"
 #include "raft.h"
 #include "roles.h"
 #include "server.h"
@@ -138,8 +139,11 @@ static void removeDomain(uint64_t domain, struct compare_data *data)
 		}
 	}
 }
-
+#if defined(__APPLE__)
+static int compareNodesForPromotion(void *p, const void *l, const void *r)
+#else
 static int compareNodesForPromotion(const void *l, const void *r, void *p)
+#endif
 {
 	struct compare_data *data = p;
 	const struct all_node_info *left = l;
@@ -165,11 +169,18 @@ static int compareNodesForPromotion(const void *l, const void *r, void *p)
 	return (left->role > right->role) - (left->role < right->role);
 }
 
+#if defined(__APPLE__)
+static int compareNodesForDemotion(void *p, const void *l, const void *r)
+{
+	return -compareNodesForPromotion(p, l, r);
+}
+#else
 static int compareNodesForDemotion(const void *l, const void *r, void *p)
 {
 	/* XXX */
 	return -compareNodesForPromotion(l, r, p);
 }
+#endif
 
 static void changeCb(struct raft_change *change, int status);
 
@@ -279,7 +290,7 @@ void RolesComputeChanges(int voters,
 
 	/* If we don't have enough voters, promote some standbys and spares. */
 	if (voter_count < voters) {
-		qsort_r(cluster, n_cluster, sizeof *cluster,
+		qsort_r_wrapper(cluster, n_cluster, sizeof *cluster,
 			compareNodesForPromotion, &voter_compare);
 	}
 	for (i = 0; i < n_cluster && voter_count < voters; i += 1) {
@@ -303,7 +314,7 @@ void RolesComputeChanges(int voters,
 	 * be picked up in the next step, and the two role changes will be
 	 * consolidated by queueChangeCb. */
 	if (voter_count > voters) {
-		qsort_r(cluster, n_cluster, sizeof *cluster,
+		qsort_r_wrapper(cluster, n_cluster, sizeof *cluster,
 			compareNodesForDemotion, &voter_compare);
 	}
 	for (i = 0; i < n_cluster && voter_count > voters; i += 1) {
@@ -318,7 +329,7 @@ void RolesComputeChanges(int voters,
 
 	/* If we don't have enough standbys, promote some spares. */
 	if (standby_count < standbys) {
-		qsort_r(cluster, n_cluster, sizeof *cluster,
+		qsort_r_wrapper(cluster, n_cluster, sizeof *cluster,
 			compareNodesForPromotion, &standby_compare);
 	}
 	for (i = 0; i < n_cluster && standby_count < standbys; i += 1) {
@@ -333,7 +344,7 @@ void RolesComputeChanges(int voters,
 
 	/* If we have too many standbys, demote some of them. */
 	if (standby_count > standbys) {
-		qsort_r(cluster, n_cluster, sizeof *cluster,
+		qsort_r_wrapper(cluster, n_cluster, sizeof *cluster,
 			compareNodesForDemotion, &standby_compare);
 	}
 	for (i = 0; i < n_cluster && standby_count > standbys; i += 1) {
@@ -604,7 +615,7 @@ static void handoverVoterCb(struct polling *polling)
 			addDomain(cluster[i].failure_domain, &voter_compare);
 		}
 	}
-	qsort_r(cluster, n_cluster, sizeof *cluster, compareNodesForPromotion,
+	qsort_r_wrapper(cluster, n_cluster, sizeof *cluster, compareNodesForPromotion,
 		&voter_compare);
 	target_id = 0;
 	for (i = 0; i < n_cluster; i += 1) {
diff --git a/src/server.c b/src/server.c
index 5910237..aba182d 100644
--- a/src/server.c
+++ b/src/server.c
@@ -115,23 +115,23 @@ int dqlite__init(struct dqlite_node *d,
 	raft_set_max_catch_up_rounds(&d->raft, 100);
 	raft_set_max_catch_up_round_duration(&d->raft, 50 * 1000); /* 50 secs */
 	raft_register_state_cb(&d->raft, state_cb);
-	rv = sem_init(&d->ready, 0, 0);
+	rv = sem_wrapper_init(&d->ready, 0);
 	if (rv != 0) {
-		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "sem_init(): %s",
+		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "semaphore_init(): %s",
 			 strerror(errno));
 		rv = DQLITE_ERROR;
 		goto err_after_raft_fsm_init;
 	}
-	rv = sem_init(&d->stopped, 0, 0);
+	rv = sem_wrapper_init(&d->stopped, 0);
 	if (rv != 0) {
-		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "sem_init(): %s",
+		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "semaphore_init(): %s",
 			 strerror(errno));
 		rv = DQLITE_ERROR;
 		goto err_after_ready_init;
 	}
-	rv = sem_init(&d->handover_done, 0, 0);
+	rv = sem_wrapper_init(&d->handover_done, 0);
 	if (rv != 0) {
-		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "sem_init(): %s",
+		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "semaphore_init(): %s",
 			 strerror(errno));
 		rv = DQLITE_ERROR;
 		goto err_after_stopped_init;
@@ -152,9 +152,9 @@ int dqlite__init(struct dqlite_node *d,
 	return 0;
 
 err_after_stopped_init:
-	sem_destroy(&d->stopped);
+	sem_wrapper_destroy(&d->stopped);
 err_after_ready_init:
-	sem_destroy(&d->ready);
+	sem_wrapper_destroy(&d->ready);
 err_after_raft_fsm_init:
 	fsm__close(&d->raft_fsm);
 err_after_raft_io_init:
@@ -178,11 +178,11 @@ void dqlite__close(struct dqlite_node *d)
 		return;
 	}
 	raft_free(d->listener);
-	rv = sem_destroy(&d->stopped);
+	rv = sem_wrapper_destroy(&d->stopped);
 	assert(rv == 0); /* Fails only if sem object is not valid */
-	rv = sem_destroy(&d->ready);
+	rv = sem_wrapper_destroy(&d->ready);
 	assert(rv == 0); /* Fails only if sem object is not valid */
-	rv = sem_destroy(&d->handover_done);
+	rv = sem_wrapper_destroy(&d->handover_done);
 	assert(rv == 0);
 	fsm__close(&d->raft_fsm);
 	// TODO assert rv of uv_loop_close after fixing cleanup logic related to
@@ -508,7 +508,7 @@ static void destroy_conn(struct conn *conn)
 static void handoverDoneCb(struct dqlite_node *d, int status)
 {
 	d->handover_status = status;
-	sem_post(&d->handover_done);
+	sem_wrapper_post(&d->handover_done);
 }
 
 static void handoverCb(uv_async_t *handover)
@@ -566,7 +566,7 @@ static void startup_cb(uv_timer_t *startup)
 	struct dqlite_node *d = startup->data;
 	int rv;
 	d->running = true;
-	rv = sem_post(&d->ready);
+	rv = sem_wrapper_post(&d->ready);
 	assert(rv == 0); /* No reason for which posting should fail */
 }
 
@@ -718,7 +718,7 @@ static int taskRun(struct dqlite_node *d)
 		snprintf(d->errmsg, DQLITE_ERRMSG_BUF_SIZE, "raft_start(): %s",
 			 raft_errmsg(&d->raft));
 		/* Unblock any client of taskReady */
-		sem_post(&d->ready);
+		sem_wrapper_post(&d->ready);
 		return rv;
 	}
 
@@ -726,7 +726,7 @@ static int taskRun(struct dqlite_node *d)
 	assert(rv == 0);
 
 	/* Unblock any client of taskReady */
-	rv = sem_post(&d->ready);
+	rv = sem_wrapper_post(&d->ready);
 	assert(rv == 0); /* no reason for which posting should fail */
 
 	return 0;
@@ -809,7 +809,7 @@ void dqlite_node_destroy(dqlite_node *d)
 static bool taskReady(struct dqlite_node *d)
 {
 	/* Wait for the ready semaphore */
-	sem_wait(&d->ready);
+	sem_wrapper_wait(&d->ready);
 	return d->running;
 }
 
@@ -921,7 +921,7 @@ int dqlite_node_handover(dqlite_node *d)
 	rv = uv_async_send(&d->handover);
 	assert(rv == 0);
 
-	sem_wait(&d->handover_done);
+	sem_wrapper_wait(&d->handover_done);
 
 	return d->handover_status;
 }
diff --git a/src/server.h b/src/server.h
index 85728aa..79250f8 100644
--- a/src/server.h
+++ b/src/server.h
@@ -3,7 +3,7 @@
 
 #include <sqlite3.h>
 
-#include <semaphore.h>
+#include "sem_wrapper.h"
 
 #include "client/protocol.h"
 #include "config.h"
@@ -31,9 +31,9 @@ struct dqlite_node {
 	struct raft_uv_transport raft_transport; /* Raft libuv transport */
 	struct raft_io raft_io;                  /* libuv I/O */
 	struct raft_fsm raft_fsm;                /* dqlite FSM */
-	sem_t ready;                             /* Server is ready */
-	sem_t stopped;                           /* Notify loop stopped */
-	sem_t handover_done;
+	sem_wrapper_t ready;                       /* Server is ready */
+	sem_wrapper_t stopped;                     /* Notify loop stopped */
+	sem_wrapper_t handover_done;
 	queue queue; /* Incoming connections */
 	queue conns; /* Active connections */
 	queue roles_changes;
diff --git a/src/tracing.c b/src/tracing.c
index a7fe351..6cdbe44 100644
--- a/src/tracing.c
+++ b/src/tracing.c
@@ -6,6 +6,9 @@
 #include <unistd.h>      /* syscall, getpid */
 #include "assert.h"      /* assert */
 #include "lib/byte.h"    /* ARRAY_SIZE */
+#if defined (__APPLE__)
+#include <pthread.h>
+#endif
 
 #define LIBDQLITE_TRACE "LIBDQLITE_TRACE"
 
@@ -52,7 +55,17 @@ static pid_t tracerPidCached;
 */
 static inline pid_t gettidImpl(void)
 {
+#ifdef __linux__
 	return (pid_t)syscall(SYS_gettid);
+#elif defined(__APPLE__)
+	uint64_t tid;
+  pthread_threadid_np(NULL, &tid);
+  return (pid_t)tid;
+#elif defined(_WIN32)
+	return (pid_t)GetCurrentThreadId();
+#else
+#error "gettidImpl is not implemented for this platform"
+#endif
 }
 
 static inline void tracerEmit(const char *file,
diff --git a/src/vfs.c b/src/vfs.c
index 062a3b4..9c9b3ac 100644
--- a/src/vfs.c
+++ b/src/vfs.c
@@ -16,7 +16,7 @@
 #include <sys/random.h>
 #include <sys/stat.h>
 #include <sys/time.h>
-#include <threads.h>
+#include "mtx_wrapper.h"
 #include <unistd.h>
 
 #include <sqlite3.h>
@@ -185,6 +185,46 @@ static bool vfsFilenameEndsWith(const char *filename, const char *suffix)
 	return strncmp(filename + n_filename - n_suffix, suffix, n_suffix) == 0;
 }
 
+/*
+ * -----------------------------------------------------------------------------
+ * Cross-platform memory remapping (mremap) workaround
+ * -----------------------------------------------------------------------------
+ *
+ * On Linux, the mremap() syscall allows remapping an existing memory mapping
+ * (such as switching between MAP_SHARED and MAP_PRIVATE) efficiently and atomically.
+ * This is used in dqlite to "shadow" or "publish" changes in shared memory regions
+ * for SQLite WAL, allowing changes to be made private to a connection or published
+ * to all connections, without SQLite being aware of the switch.
+ *
+ * However, mremap() is not available on macOS, BSD, or Windows. To provide
+ * portability, we define a portable_mremap() function for non-Linux systems that
+ * simulates mremap by unmapping the target region and mapping a new region at the
+ * same address. This is not atomic and may introduce race conditions if other
+ * threads or processes access the region during the remapping window.
+ *
+ * Limitations:
+ * - On non-Linux systems, the remapping is not atomic and may not be safe for
+ *   concurrent access.
+ * - The semantics of MAP_PRIVATE and MAP_SHARED may differ between platforms.
+ * - This workaround is only suitable for single-process or carefully synchronized
+ *   scenarios outside Linux.
+ *
+ * If you need full safety and performance, use Linux. On other systems, be aware
+ * of the limitations and test accordingly.
+ *
+ * See also: vfsRedirectShm, vfsPublishShm, vfsRollbackShm.
+ */
+#if !defined(__linux__)
+  static void *portable_mremap(void *old_address, size_t old_size, size_t new_size, int flags, void *new_address) {
+    (void)flags;
+		munmap(old_address, old_size);
+      return mmap(new_address, new_size, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_PRIVATE, -1, 0);
+  }
+  #define mremap portable_mremap
+  #define MREMAP_MAYMOVE 0
+  #define MREMAP_FIXED 0
+#endif
+
 /******************************************************************************/
 /*                            Main data structures                            */
 /******************************************************************************/
@@ -262,22 +302,61 @@ static void vfsFrameDestroy(struct vfsFrame *f)
 /* Hold content for a shared memory mapping. */
 struct vfsShm
 {
-	mtx_t mtx; /* Mutex to protect fields below. */
+	mtx_wrapper_t mtx; /* Mutex to protect fields below. */
 	int fd;     /* Read-only */
 	int size;   /* Total size of the in-memory file */
 	/* Lock array. Each of these lock has the following semantics:
 	 *  -  0 means unlocked;
 	 *  - -1 means exclusive locked;
-	 *  - >0 means shared locked and the value is the count of 
+	 *  - >0 means shared locked and the value is the count of
 	 *       shared locks taken.
 	 */
 	int lock[SQLITE_SHM_NLOCK];
 };
 
+/* Cross-platform memfd_create replacement */
+static int portable_memfd_create(const char *name) {
+#if defined(__linux__)
+  return memfd_create(name, 0);
+#elif defined(__APPLE__) || defined(__unix__)
+  char tmpname[256];
+  int fd;
+  snprintf(tmpname, sizeof(tmpname), "/tmp/%s-XXXXXX", name ? name : "dqlite-shm");
+  fd = mkstemp(tmpname);
+  if (fd < 0) {
+      return -1;
+  }
+  /* Unlink so it is removed after close */
+  unlink(tmpname);
+  return fd;
+#elif defined(_WIN32)
+	char tempFile[MAX_PATH];
+	if (!GetTempPathA(MAX_PATH, tempFile)) {
+	    return -1;
+	}
+	if (!GetTempFileNameA(tempFile, name ? name : "dqlite-shm", 0, tempFile)) {
+	    return -1;
+	}
+	HANDLE hFile = CreateFileA(
+	    tempFile,
+	    GENERIC_READ | GENERIC_WRITE,
+	    0, NULL,
+	    CREATE_ALWAYS,
+	    FILE_ATTRIBUTE_TEMPORARY | FILE_FLAG_DELETE_ON_CLOSE,
+	    NULL
+	);
+	if (hFile == INVALID_HANDLE_VALUE) {
+	  return -1;
+	}
+	int fd = _open_osfhandle((intptr_t)hFile, _O_RDWR);
+	return fd;
+#endif
+}
+
 /* Initialize the shared memory mapping of a database file. */
 static int vfsShmInit(struct vfsShm *s)
 {
-	int fd = memfd_create("dqlite-shm", 0);
+	int fd = portable_memfd_create("dqlite-shm");
 	if (fd < 0) {
 		return SQLITE_NOMEM;
 	}
@@ -291,13 +370,13 @@ static int vfsShmLock(struct vfsShm *s, int ofst, int n, bool exclusive)
 {
 	PRE(ofst >= 0 && n > 0 && ofst + n <= SQLITE_SHM_NLOCK);
 
-	mtx_lock(&s->mtx);
+	mtx_wrapper_lock(&s->mtx);
 	for (int i = ofst; i < ofst + n; i++) {
 		if (exclusive && s->lock[i] != 0) {
-			mtx_unlock(&s->mtx);
+			mtx_wrapper_unlock(&s->mtx);
 			return SQLITE_BUSY;
 		} else if (!exclusive && s->lock[i] < 0) {
-			mtx_unlock(&s->mtx);
+			mtx_wrapper_unlock(&s->mtx);
 			return SQLITE_BUSY;
 		}
 	}
@@ -310,7 +389,7 @@ static int vfsShmLock(struct vfsShm *s, int ofst, int n, bool exclusive)
 		}
 	}
 
-	mtx_unlock(&s->mtx);
+	mtx_wrapper_unlock(&s->mtx);
 	return SQLITE_OK;
 }
 
@@ -318,7 +397,7 @@ static int vfsShmUnlock(struct vfsShm *s, int ofst, int n, bool exclusive)
 {
 	PRE(ofst >= 0 && n > 0 && ofst + n <= SQLITE_SHM_NLOCK);
 
-	mtx_lock(&s->mtx);
+	mtx_wrapper_lock(&s->mtx);
 	for (int i = ofst; i < ofst + n; i++) {
 		if (exclusive) {
 			PRE(s->lock[i] < 0);
@@ -328,7 +407,7 @@ static int vfsShmUnlock(struct vfsShm *s, int ofst, int n, bool exclusive)
 			s->lock[i]--;
 		}
 	}
-	mtx_unlock(&s->mtx);
+	mtx_wrapper_unlock(&s->mtx);
 	return SQLITE_OK;
 }
 
@@ -353,7 +432,7 @@ struct vfsWal
 	struct vfsFrame **tx; /* Frames added by a transaction. */
 	unsigned n_tx;        /* Number of added frames. */
 
-	mtx_t mtx;                /* Lock for fields below. */
+	mtx_wrapper_t mtx;                /* Lock for fields below. */
 	struct vfsFrame **frames; /* All frames committed. */
 	unsigned n_frames;        /* Number of committed frames. */
 };
@@ -362,7 +441,7 @@ struct vfsWal
 static void vfsWalInit(struct vfsWal *w)
 {
 	*w = (struct vfsWal){};
-	int rv = mtx_init(&w->mtx, mtx_plain);
+	int rv = mtx_wrapper_init(&w->mtx);
 	assert(rv == 0);
 }
 
@@ -374,16 +453,16 @@ static struct vfsFrame *vfsWalFrameLookup(struct vfsWal *w, unsigned n)
 	assert(w != NULL);
 	assert(n > 0);
 
-	mtx_lock(&w->mtx);
+	mtx_wrapper_lock(&w->mtx);
 	if (n <= w->n_frames) {
 		frame = w->frames[n - 1];
 	} else {
 		frame = NULL;
 	}
-	mtx_unlock(&w->mtx);
+	mtx_wrapper_unlock(&w->mtx);
 
 	/* In case of a cache spill, it is possible that SQLite will try to
-	 * re-populate the cache reading from the uncommitted part of the WAL. 
+	 * re-populate the cache reading from the uncommitted part of the WAL.
 	 * Unfortunately, it is not possible to reliably link a WAL file to
 	 * its main file and checking this way that this is the connection
 	 * holding the write lock. As such, the best we can do is to trust
@@ -410,7 +489,7 @@ static uint32_t vfsWalGetPageSize(struct vfsWal *w)
 /* Return the size of the WAL file in bytes. */
 static int64_t vfsWalSize(struct vfsWal *w)
 {
-	mtx_lock(&w->mtx);
+	mtx_wrapper_lock(&w->mtx);
 	int64_t size = 0;
 	if (w->n_frames > 0) {
 		uint32_t page_size;
@@ -419,7 +498,7 @@ static int64_t vfsWalSize(struct vfsWal *w)
 		size += (int64_t)w->n_frames *
 			(int64_t)(FORMAT__WAL_FRAME_HDR_SIZE + page_size);
 	}
-	mtx_unlock(&w->mtx);
+	mtx_wrapper_unlock(&w->mtx);
 	/* TODO dqlite is limited to a max database size of SIZE_MAX */
 	assert(size >= 0 && (uint64_t)size <= SIZE_MAX);
 	return size;
@@ -427,7 +506,7 @@ static int64_t vfsWalSize(struct vfsWal *w)
 
 static void vfsWalTruncate(struct vfsWal *w)
 {
-	mtx_lock(&w->mtx);
+	mtx_wrapper_lock(&w->mtx);
 	PRE(w->n_tx == 0);
 	PRE(w->tx == NULL);
 	for (unsigned i = 0; i < w->n_frames; i++) {
@@ -437,7 +516,7 @@ static void vfsWalTruncate(struct vfsWal *w)
 
 	w->frames = NULL;
 	w->n_frames = 0;
-	mtx_unlock(&w->mtx);
+	mtx_wrapper_unlock(&w->mtx);
 }
 
 /* Release all memory used by a WAL object. */
@@ -451,7 +530,7 @@ static void vfsWalClose(struct vfsWal *w)
 	w->n_tx = 0;
 
 	vfsWalTruncate(w);
-	mtx_destroy(&w->mtx);
+	mtx_wrapper_destroy(&w->mtx);
 }
 
 /* Database-specific content */
@@ -462,7 +541,7 @@ struct vfsDatabase
 	struct vfsShm shm;  /* Shared memory. */
 	struct vfsWal wal;  /* Associated WAL. */
 
-	mtx_t mtx;
+	mtx_wrapper_t mtx;
 	void **pages;       /* All database. */
 	unsigned n_pages;   /* Number of pages. */
 };
@@ -511,7 +590,7 @@ static int vfsDatabaseInit(struct vfsDatabase *d, const char *name)
 		return rv;
 	}
 	vfsWalInit(&d->wal);
-	rv = mtx_init(&d->mtx, mtx_plain);
+	rv = mtx_wrapper_init(&d->mtx);
 	assert(rv == 0);
 	return SQLITE_OK;
 }
@@ -554,7 +633,7 @@ static int vfsDatabaseGetPage(struct vfsDatabase *d,
 		goto err;
 	}
 
-	mtx_lock(&d->mtx);
+	mtx_wrapper_lock(&d->mtx);
 	void **pages = sqlite3_realloc64(d->pages, sizeof *pages * pgno);
 	if (pages == NULL) {
 		rc = SQLITE_NOMEM;
@@ -576,14 +655,14 @@ static int vfsDatabaseGetPage(struct vfsDatabase *d,
 	/* Update the page array. */
 	d->pages = pages;
 	d->n_pages = pgno;
-	mtx_unlock(&d->mtx);
+	mtx_wrapper_unlock(&d->mtx);
 	return SQLITE_OK;
 
 err_after_pending_byte_page:
 	d->pages = pages;
 
 err_after_vfs_page_create:
-	mtx_unlock(&d->mtx);
+	mtx_wrapper_unlock(&d->mtx);
 	sqlite3_free(*page);
 err:
 	*page = NULL;
@@ -622,10 +701,10 @@ static uint32_t vfsDatabaseGetPageSize(struct vfsDatabase *d)
 		return d->page_size;
 	}
 
-	mtx_lock(&d->mtx);
+	mtx_wrapper_lock(&d->mtx);
 	assert(d->n_pages > 0);
 	page = d->pages[0];
-	mtx_unlock(&d->mtx);
+	mtx_wrapper_unlock(&d->mtx);
 
 	/* The page size is stored in the 16th and 17th bytes of the first
 	 * database page (big-endian) */
@@ -635,9 +714,9 @@ static uint32_t vfsDatabaseGetPageSize(struct vfsDatabase *d)
 /* Return the size of the database file in bytes. */
 static int64_t vfsDatabaseFileSize(struct vfsDatabase *d)
 {
-	mtx_lock(&d->mtx);
+	mtx_wrapper_lock(&d->mtx);
 	int64_t pages = (int64_t)d->n_pages;
-	mtx_unlock(&d->mtx);
+	mtx_wrapper_unlock(&d->mtx);
 
 	if (pages == 0) {
 		return 0;
@@ -674,7 +753,7 @@ static int vfsDatabaseTruncate(struct vfsDatabase *d, sqlite_int64 size)
 	assert(n_pages <= d->n_pages);
 	assert(d->pages != NULL);
 
-	mtx_lock(&d->mtx);
+	mtx_wrapper_lock(&d->mtx);
 	/* Destroy pages beyond pages_len. */
 	void **cursor = d->pages + n_pages;
 	for (unsigned i = 0; i < (d->n_pages - n_pages); i++) {
@@ -693,7 +772,7 @@ static int vfsDatabaseTruncate(struct vfsDatabase *d, sqlite_int64 size)
 
 	/* Update the page count. */
 	d->n_pages = n_pages;
-	mtx_unlock(&d->mtx);
+	mtx_wrapper_unlock(&d->mtx);
 
 	return SQLITE_OK;
 }
@@ -708,7 +787,7 @@ static void vfsDatabaseClose(struct vfsDatabase *d)
 	sqlite3_free(d->name);
 	vfsWalClose(&d->wal);
 	vfsShmClose(&d->shm);
-	mtx_destroy(&d->mtx);
+	mtx_wrapper_destroy(&d->mtx);
 }
 
 /* Custom dqlite VFS. Contains pointers to all databases that were created. */
@@ -1098,7 +1177,7 @@ static int vfsWalFileWrite(sqlite3_file* file, const void* buf, int amount, sqli
 	page_size = vfsWalGetPageSize(&f->database->wal);
 	assert(page_size > 0);
 
-	/* It is expected that once committed the WAL cannot be changed. 
+	/* It is expected that once committed the WAL cannot be changed.
 	 * As such, its index must be above the already committed part */
 	/* FIXME: this should likely be called "pageno" as it's 1-based. */
 	index = (unsigned)formatWalCalcFrameIndex((int)page_size, offset);
@@ -1319,7 +1398,7 @@ static int vfsMainFileWrite(sqlite3_file *file,
 	 *   Moreover, this logic will always run in the libuv thread.
 	 *
 	 * The above reasoning would look like synchronization through mutexes is not
-	 * strictly needed here, however that is only partially true: SQLite might 
+	 * strictly needed here, however that is only partially true: SQLite might
 	 * use the `xFileSize` (vfsMainFileSize) to check for file existance without
 	 * taking any lock first (i.e. the only locking guarantee is on read and write).
 	 * As such, the code below will still use a (likely uncontended) mutex to guard
@@ -1485,23 +1564,23 @@ static int vfsMainFileShmMap(sqlite3_file *file, /* Handle open on database file
 	PRE(region_size == VFS__WAL_INDEX_REGION_SIZE);
 
 	int rv;
-	mtx_lock(&s->mtx);
+	mtx_wrapper_lock(&s->mtx);
 	if (s->size < offset + region_size) {
 		if (extend) {
 			/* The file is not big enough */
 			rv = ftruncate(s->fd, offset + region_size);
 			if (rv < 0) {
-				mtx_unlock(&s->mtx);
+				mtx_wrapper_unlock(&s->mtx);
 				return SQLITE_NOMEM;
 			}
 			s->size = offset + region_size;
 		} else {
 			*out = NULL;
-			mtx_unlock(&s->mtx);
+			mtx_wrapper_unlock(&s->mtx);
 			return SQLITE_OK;
 		}
 	}
-	mtx_unlock(&s->mtx);
+	mtx_wrapper_unlock(&s->mtx);
 
 	while (f->mappedShmRegions.cap <= region_index) {
 		int newCap = 2 * f->mappedShmRegions.cap + 1;
@@ -1578,7 +1657,7 @@ static void vfsWalRollbackIfUncommitted(struct vfsWal *w)
  *
  * See also vfsPublishShm and vfsRollbackShm. */
 static int vfsRedirectShm(struct vfsMainFile *f) {
-	/* When taking a write lock also make all mappings private. 
+	/* When taking a write lock also make all mappings private.
 	 * This effectively shadows the transaction as all other
 	 * connections cannot see the update to the mxFrames field. */
 	for (int regionIndex = 0; regionIndex < f->mappedShmRegions.len; regionIndex++) {
@@ -1586,12 +1665,12 @@ static int vfsRedirectShm(struct vfsMainFile *f) {
 		void *new_region = mmap(NULL, VFS__WAL_INDEX_REGION_SIZE,
 		PROT_READ | PROT_WRITE, MAP_PRIVATE, f->database->shm.fd, regionIndex * VFS__WAL_INDEX_REGION_SIZE);
 		if (new_region == MAP_FAILED) {
-			/* This should never happen. Also, this means that we might leave the 
+			/* This should never happen. Also, this means that we might leave the
 			* connection in a weird state. */
 			return SQLITE_NOMEM;
 		}
-		void *remapped = mremap(new_region, 
-			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE, 
+		void *remapped = mremap(new_region,
+			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE,
 			MREMAP_MAYMOVE | MREMAP_FIXED, region);
 		assert(remapped == region);
 	}
@@ -1611,7 +1690,7 @@ static int vfsPublishShm(struct vfsMainFile *f) {
 	}
 
 	/* The shared memory needs to be written in the proper order so that we don't end up
-	 * with concurrency issues. 
+	 * with concurrency issues.
 	 * It is possible to split the shared memory in 2 parts:
 	 *  - the header, composed by
 	 *    - two copies actually of the WAL Index Information section
@@ -1627,8 +1706,8 @@ static int vfsPublishShm(struct vfsMainFile *f) {
 			PROT_READ | PROT_WRITE, MAP_SHARED, f->database->shm.fd, i * VFS__WAL_INDEX_REGION_SIZE);
 		assert(region != MAP_FAILED);
 		memcpy(region, f->mappedShmRegions.ptr[i], VFS__WAL_INDEX_REGION_SIZE);
-		void *remapped = mremap(region, 
-			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE, 
+		void *remapped = mremap(region,
+			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE,
 			MREMAP_MAYMOVE | MREMAP_FIXED, f->mappedShmRegions.ptr[i]);
 		assert(remapped == f->mappedShmRegions.ptr[i]);
 	}
@@ -1652,14 +1731,14 @@ static int vfsPublishShm(struct vfsMainFile *f) {
 	memcpy(first_region_shared, f->mappedShmRegions.ptr[0], VFS__WAL_INDEX_HEADER_SIZE);
 
 	/* Check that no checkpointing happened during the write transaction. */
-	
+
 	/* Now it is necessary to merge the synchronization part.
 	 * Read marks are never changed by a write transaction as a write transaction can
 	 * only be started after a read transaction and as such the marker and the read lock is
 	 * already acquired.
 	 * For the other fields, quoting the documentation:
 	 *  > The nBackfill can only be increased while holding the WAL_CKPT_LOCK. However, nBackfill
-	 *  > is changed to zero during a WAL reset, and this happens while holding the WAL_WRITE_LOCK. 
+	 *  > is changed to zero during a WAL reset, and this happens while holding the WAL_WRITE_LOCK.
 	 * And also:`
 	 *  > The mxFrame value is always greater than or equal to both nBackfill and nBackfillAttempted.
 	 * Which means that only when holding the checkpoint lock those fields should be published */
@@ -1679,7 +1758,7 @@ static int vfsPublishShm(struct vfsMainFile *f) {
 	}
 
 	void *remapped = mremap(first_region_shared,
-		VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE, 
+		VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE,
 		MREMAP_MAYMOVE | MREMAP_FIXED, f->mappedShmRegions.ptr[0]);
 	assert(remapped == f->mappedShmRegions.ptr[0]);
 
@@ -1695,13 +1774,13 @@ static int vfsRollbackShm(struct vfsMainFile *f) {
 		void *new_region = mmap(NULL, VFS__WAL_INDEX_REGION_SIZE,
 			PROT_READ | PROT_WRITE, MAP_SHARED, f->database->shm.fd, i * VFS__WAL_INDEX_REGION_SIZE);
 		if (new_region == MAP_FAILED) {
-			/* This should never happen. Also, this means that we might leave the 
+			/* This should never happen. Also, this means that we might leave the
 				* connection in a weird state. */
 			return SQLITE_NOMEM;
 		}
 		/* Now the private map can be unmapped safely going back to the shared one. */
-		void *remapped = mremap(new_region, 
-			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE, 
+		void *remapped = mremap(new_region,
+			VFS__WAL_INDEX_REGION_SIZE, VFS__WAL_INDEX_REGION_SIZE,
 			MREMAP_MAYMOVE | MREMAP_FIXED, f->mappedShmRegions.ptr[i]);
 		assert(remapped == f->mappedShmRegions.ptr[i]);
 	}
@@ -1910,7 +1989,7 @@ static int vfsDiskFileWrite(sqlite3_file *file,
 			    sqlite_int64 offset)
 {
 	struct vfsDiskMainFile *f = (struct vfsDiskMainFile *)file;
-	
+
 	/* Write to the actual database file. */
 	vfsDiskDatabaseTrackNumPages(f->base.database, offset);
 	int rv = f->underlying->pMethods->xWrite(f->underlying, buf, amount, offset);
@@ -2300,13 +2379,27 @@ static void vfsDlClose(sqlite3_vfs *vfs, void *pHandle)
 static int vfsRandomness(sqlite3_vfs *vfs, int nByte, char *zByte)
 {
 	(void)vfs;
-
+#if defined(__linux__)
 	ssize_t rv = getrandom(zByte, (size_t)nByte, 0);
 	if (rv == -1) {
 		/* Ignore failed attempts */
 		return 0;
 	}
 	return (int)rv;
+#elif defined(__APPLE__)
+  arc4random_buf(zByte, (size_t)nByte);
+  return nByte;
+#elif defined(_WIN32)
+  HCRYPTPROV hProv;
+  if (CryptAcquireContext(&hProv, NULL, NULL, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT)) {
+      BOOL ok = CryptGenRandom(hProv, nByte, (BYTE*)zByte);
+      CryptReleaseContext(hProv, 0);
+      return ok ? nByte : 0;
+  }
+  return 0;
+#else
+#error "No vfsRandomness implementation for this platform"
+#endif
 }
 
 static int vfsSleep(sqlite3_vfs *vfs, int microseconds)
@@ -2508,15 +2601,15 @@ static int vfsWalAppend(struct vfsDatabase *d,
 		database_size = vfsFrameGetDatabaseSize(frame);
 	}
 
-	mtx_lock(&w->mtx);
+	mtx_wrapper_lock(&w->mtx);
 	frames = sqlite3_realloc64(
 	    w->frames, sizeof(*frames) * (w->n_frames + transaction->n_pages));
 	if (frames == NULL) {
-		mtx_unlock(&w->mtx);
+		mtx_wrapper_unlock(&w->mtx);
 		goto oom;
 	}
 	w->frames = frames;
-	mtx_unlock(&w->mtx);
+	mtx_wrapper_unlock(&w->mtx);
 
 	for (i = 0; i < transaction->n_pages; i++) {
 		struct vfsFrame *frame = vfsFrameCreate(page_size);
@@ -2528,7 +2621,7 @@ static int vfsWalAppend(struct vfsDatabase *d,
 			goto oom_after_frames_alloc;
 		}
 
-		/* When writing the SQLite database header, make sure to sync 
+		/* When writing the SQLite database header, make sure to sync
 		 * the file size to the logical database size. */
 		if (page_number == 1) {
 			database_size = ByteGetBe32(&page[VFS__IN_HEADER_DATABASE_SIZE_OFFSET]);
@@ -2546,9 +2639,9 @@ static int vfsWalAppend(struct vfsDatabase *d,
 		frames[w->n_frames + i] = frame;
 	}
 
-	mtx_lock(&w->mtx);
+	mtx_wrapper_lock(&w->mtx);
 	w->n_frames += transaction->n_pages;
-	mtx_unlock(&w->mtx);
+	mtx_wrapper_unlock(&w->mtx);
 
 	return 0;
 
diff --git a/test/integration/test_node.c b/test/integration/test_node.c
index d67a25a..7fad3ff 100644
--- a/test/integration/test_node.c
+++ b/test/integration/test_node.c
@@ -1,4 +1,4 @@
-#include <semaphore.h>
+#include "../../src/sem_wrapper.h"
 #include <time.h>
 #include <unistd.h>
 #include "../lib/fs.h"
@@ -234,7 +234,7 @@ TEST(node, stopInflightReads, setUpInet, tearDown, 0, node_params)
 
 		rv = openDb(&clients[i], f->node, "test");
 		munit_assert_int(rv, ==, RAFT_OK);
-		rv = clientSendQuerySQL(&clients[i], 
+		rv = clientSendQuerySQL(&clients[i],
 			"WITH RECURSIVE inf(i) AS( "
 			"    SELECT 1              "
 			"	UNION ALL              "
@@ -244,7 +244,7 @@ TEST(node, stopInflightReads, setUpInet, tearDown, 0, node_params)
 			NULL, 0, NULL
 		);
 		munit_assert_int(rv, ==, RAFT_OK);
-		
+
 		rv = clientRecvRows(&clients[i], &rows, &done, NULL);
 		munit_assert_int(rv, ==, RAFT_OK);
 		munit_assert_false(done);
@@ -273,21 +273,21 @@ static int register_notify(sqlite3 *connection, char **pzErrMsg, struct sqlite3_
 					  NULL, NULL, NULL);
 }
 
-static sem_t write_sem;
+static sem_wrapper_t write_sem;
 static void notify_transaction(sqlite3_context *context, int argc, sqlite3_value **argv)
 {
 	(void)argc;
 	/* Just return the same value */
 	sqlite3_result_value(context, argv[0]);
 	/* Signal the application that we were able to get to the write part */
-	sem_post(&write_sem);
+	sem_wrapper_post(&write_sem);
 }
 
 static void *setUpInflight(const MunitParameter params[], void *user_data)
 {
 	void *fixture = setUpInet(params, user_data);;
 
-	int rv = sem_init(&write_sem, 0, 0);
+	int rv = sem_wrapper_init(&write_sem, 0);
 	munit_assert_int(rv, ==, 0);
 
 	rv = sqlite3_auto_extension((void (*)(void))register_notify);
@@ -301,7 +301,7 @@ static void tearDownInflight(void *data)
 	int rv = sqlite3_cancel_auto_extension((void (*)(void))notify_transaction);
 	munit_assert_int(rv, ==, SQLITE_OK);
 
-	rv = sem_destroy(&write_sem);
+	rv = sem_wrapper_destroy(&write_sem);
 	munit_assert_int(rv, ==, 0);
 
 	tearDown(data);
@@ -334,12 +334,12 @@ TEST(node, stopInflightWrites, setUpInflight, tearDownInflight, 0, node_params)
 	munit_assert_int(started, ==, 0);
 
 	for (int i = 0; i < CLIENT_N; i++) {
-		rv = clientSendExecSQL(&clients[i], 
+		rv = clientSendExecSQL(&clients[i],
 			"INSERT INTO test_table VALUES (notify_transaction(1))", NULL, 0, NULL);
 		munit_assert_int(rv, ==, RAFT_OK);
 	}
 
-	sem_wait(&write_sem);
+	sem_wrapper_wait(&write_sem);
 
 	/* Make sure the node can be closed */
 	rv = dqlite_node_stop(f->node);

```
</details>


I think it would be very important to have cross-platform support, at least for Windows and Linux (Android is Linux).
If I can help contribute, please let me know.
Thanks
